tools/schemapi/schemapi.py

```diff
@@ -170,14 +225,41 @@ def __str__(self):
             message += "\n        " + "\n        ".join(
                 [e.message for e in self._additional_errors]
             )
-        return """Invalid specification
-
-        {}, validating {!r}
 
-        {}
-        """.format(
-            schema_path, self.validator, message
-        )
+        if hasattr(vegalite, schema_path.split(".")[-1]):
```

Comment:
@binste I noticed an issue here after #2842 was merged. Previously, error messages would include a reference to the class where the error was raised, as in this case with the Y channel:

```
altair.vegalite.v5.schema.channels.Y, validating 'additionalProperties'
```

After #2842 the first part of that error message changes to no longer point to the particular class where the error was encountered, but just the chart in general:

```
altair.vegalite.v5.api.Chart->0, validating 'additionalProperties'
```

In addition to being less informative in general, this becomes an issue here because this PR was relying on that first part of the error message to find all the correct parameters names that the Y channel accept. There might be ways to work around that in this PR, but I wanted to ask you first if you think it is possible to restore the more precise error message that points to the class where the error was encountered or if that information is lost with the new approach to raising errors.

Reply:
To add some more info, previously with deep validation, what I think was happening was that the iterative finding of the error meant that the class of the SchemaBase (I think) would change in each iteration to reflect which part of the spec was checked for errors. For example if I print out `type(self)` of the `self` that is passed to the `SchemaValidationError` it would look like this with deep validation (for a spec with a typo in a param in the Y channel):

```
<class 'altair.vegalite.v5.api.Chart'>
<class 'altair.vegalite.v5.schema.channels.Y'>
```
Without deep validation, all the errors are found without iteration and therefor the `self` param remains as the chart itself without changing to the Y channel where the error is occurring:

```
<class 'altair.vegalite.v5.api.Chart'>
```

I tried to look briefly if there was a way to not only return all the error messages, but also all the classes so that we could still find out that in this case it was the Y class that the error pertained to.

I looked briefly and it seems like some of this information is stored in the `absolute_path` attribute of the `self` object inside the `SchemaValidationError` class, and it might be possible to reconstruct it from there, but I couldn't find the full class path anywhere and I wanted to hear your thoughts before going down that path.

Reply:
Thanks for the detailed write-up! Indeed, I agree that it would be more informative if that original behaviour is restored and of course helpful for your proposed changes here, which look very helpful.

I think `absolute_path` is the correct place to take this from. From that name, we can try to use the usual conversion to a Python name that is used in the propertysetter https://github.com/altair-viz/altair/blob/master/altair/utils/schemapi.py#L680 and also in the code generation https://github.com/altair-viz/altair/blob/master/tools/generate_schema_wrapper.py#L454
```python
classname = prop[0].upper() + prop[1:]
```
and then fetch that class from `vegalite`. So in `SchemaValidationError.__str__` instead of `cls = self.obj.__class__` we could do something like

```python
class_name_lower = self.absolute_path[-1]
class_name = f"{class_name_lower}"[0].upper() + f"{class_name_lower}"[1:]
cls = getattr(vegalite, class_name)
cls
```
and before that in `SchemaValidationError.__init__`:

```python
self.absolute_path = err.absolute_path
```

I haven't tested this yet but most classes should be importable from the top-level like this right? For others we could still fall back on `self.obj` (the `self` that is passed into `SchemaValidationError`).

If I find more time in the next days I can make a PR with this to restore the original behaviour of showing the class in which the error occured.

Reply:
See #2883

Reply:
Thank you! I merged in your PR and this seems to be working as intended now. There are a few failing tests which needs the error message updated; I think I can get to those on Friday or Sunday this week. There seems to be some larger error also where my solution here still needs some change to work with the new multiple returned error messages.